pandas/core/arrays/base.py

```diff
@@ -0,0 +1,191 @@
+"""An interface for extending pandas with custom arrays."""
+import abc
+
+import numpy as np
+
+from pandas.compat import add_metaclass
+
+
+_not_implemented_message = "{} does not implement {}."
+
+
+@add_metaclass(abc.ABCMeta)
+class ExtensionArray(object):
+    """Abstract base class for custom array types
+
+    Notes
+    -----
+    pandas will recognize instances of this class as proper arrays
+    with a custom type and will not attempt to coerce them to objects.
+
+    **Restrictions on your class constructor**
+
+        * Extension arrays should be able to be constructed with instances of
+          the class, i.e. ``ExtensionArray(extension_array)`` should return
+          an instance, not error.
+    """
+    # ------------------------------------------------------------------------
+    # Must be a Sequence
+    # ------------------------------------------------------------------------
+    @abc.abstractmethod
+    def __getitem__(self, item):
+        # type (Any) -> Any
+        """Select a subset of self.
+
+        Notes
+        -----
+        ``item`` may be one of
+
+            * A scalar integer position
+            * A slice object, where 'start', 'stop', and 'step' are
+              integers or None
+            * A 1-d boolean NumPy ndarray the same length as 'self'
```

Comment:
We should be strict about how we use this interface from pandas. I see little harm in saying this is always a NumPy array since we are the only ones calling it and some clear advantages. Otherwise, how is a user supposed to detect this case?